optimization.ts

```diff
@@ -0,0 +1,433 @@
+    let preds: Map<string, string[]> = new Map<string, string[]>();
+    let blockMapping: Map<string, BasicBlock<any>> = new Map<string, BasicBlock<any>>();
+    basicBlocks.forEach(basicBlock=>{
+        blockMapping.set(basicBlock.label, basicBlock);
+        const lastStmt = basicBlock.stmts[basicBlock.stmts.length-1];
+        if(lastStmt !== undefined && lastStmt.tag === "ifjmp"){
+            //Assigning successors
+            if (succs.has(basicBlock.label) && !succs.get(basicBlock.label).includes(lastStmt.thn))
+                succs.set(basicBlock.label, [...succs.get(basicBlock.label), lastStmt.thn]);
+            else if (!succs.has(basicBlock.label))
+                succs.set(basicBlock.label, [lastStmt.thn]);
+            
+
+            if (succs.has(basicBlock.label) && !succs.get(basicBlock.label).includes(lastStmt.els))
+                succs.set(basicBlock.label, [...succs.get(basicBlock.label), lastStmt.els]);
+            else if (!succs.has(basicBlock.label))
+                succs.set(basicBlock.label, [lastStmt.els]);
+
+            //Assigning predecessors
+            if(preds.has(lastStmt.thn) && !preds.get(lastStmt.thn).includes(basicBlock.label))
+                preds.set(lastStmt.thn, [...preds.get(lastStmt.thn), basicBlock.label]);
+            else if (!preds.has(lastStmt.thn))
+                preds.set(lastStmt.thn, [basicBlock.label]);
+
+            if(preds.has(lastStmt.els) && !preds.get(lastStmt.els).includes(basicBlock.label))
+                preds.set(lastStmt.els, [...preds.get(lastStmt.els), basicBlock.label]);
+            else if (!preds.has(lastStmt.els))
+                preds.set(lastStmt.els, [basicBlock.label]);
+        }
+        else if (lastStmt !== undefined && lastStmt.tag === "jmp"){
+            //Assigning successors
+            if (succs.has(basicBlock.label) && !succs.get(basicBlock.label).includes(lastStmt.lbl))
+                succs.set(basicBlock.label, [...succs.get(basicBlock.label), lastStmt.lbl]);
+            else if (!succs.has(basicBlock.label))
+                succs.set(basicBlock.label, [lastStmt.lbl]);
+
+            //Assigning predecessors
+            if (preds.has(lastStmt.lbl) && !preds.get(lastStmt.lbl).includes(basicBlock.label))
+                preds.set(lastStmt.lbl, [...preds.get(lastStmt.lbl), basicBlock.label]);
+            else if (!preds.has(lastStmt.lbl))
+                preds.set(lastStmt.lbl, [basicBlock.label]);
+        }
+    });
+    return [preds, succs, blockMapping];
+
+}
+
+function computeInitEnv(varDefs: Array<VarInit<any>>, dummyEnv: boolean): Env{
+    var env: Env = {vars: new Map<string, compileVal>()};
+    varDefs.forEach(def => {
+        if (!dummyEnv)
+            env.vars.set(def.name, {tag: "val", value: def.value});
+        else
+            env.vars.set(def.name, {tag: "undef"});
+    });
+    return env;
+}
+
+function mergeEnvironment(a: Env, b: Env): Env{
```

Comment:
This kind of thing is hard for me to review accurately + quickly.

The main thing I want to give as feedback is that I think you need many, many more tests, given the complexity of this code. For example, none of the tests even have >=3 variables to check if merging interesting environments could be an issue, that's just off the top of my head.

I'd like to see a bunch of tests written with the goal of _breaking_ this code.